src/tribol/search/InterfacePairFinder.cpp

```diff
@@ -1,643 +1,873 @@
+   const int dim = m_coupling_scheme->spatialDimension();
+   axom::slic::flushStreams();
+   m_search = nullptr;
 
-   if( m_gridSearch3D != nullptr)
+   if (isOnDevice(cs->getExecutionMode()) && cs->getBinningMethod() == BINNING_GRID)
    {
-      delete m_gridSearch3D;
-      m_gridSearch3D = nullptr;
+      SLIC_WARNING_ROOT("BINNING_GRID is not supported on GPU. Switching to BINNING_BVH.");
+      cs->setBinningMethod(BINNING_BVH);
    }
 
-}
-
-void InterfacePairFinder::initialize()
-{
-   const int dim = m_couplingScheme->spatialDimension();
-
-   switch(m_couplingScheme->getBinningMethod() )
+   switch(cs->getBinningMethod() )
    {
    case BINNING_CARTESIAN_PRODUCT:
-      // no-op
+      switch( dim )
+      {
+      case 2:
+         m_search = new CartesianProduct<2>(m_coupling_scheme);
+         break;
+      case 3:
+         m_search = new CartesianProduct<3>(m_coupling_scheme);
+         break;
+      default:
+         SLIC_ERROR("Invalid dimension: " << dim );
+         break;
+      } // end of BINNING_CARTESIAN_PRODUCT dimension switch
       break;
    case BINNING_GRID:
       // The spatial grid is templated on the dimension
       switch( dim )
       {
       case 2:
-         m_gridSearch2D = new GridSearch<2>(m_couplingScheme);
-         m_gridSearch2D->generateSpatialIndex();
+         m_search = new GridSearch<2>(m_coupling_scheme);
          break;
       case 3:
-         m_gridSearch3D = new GridSearch<3>(m_couplingScheme);
-         m_gridSearch3D->generateSpatialIndex();
+         m_search = new GridSearch<3>(m_coupling_scheme);
          break;
       default:
          SLIC_ERROR("Invalid dimension: " << dim );
          break;
-      }
+      } // end of BINNING_GRID dimension switch
       break;
-   default:
-      SLIC_ERROR("Unsupported binning method");
-      break;
-   }
-}
-
-void InterfacePairFinder::findInterfacePairs()
-{
-   const int dim = m_couplingScheme->spatialDimension();
-
-   switch(m_couplingScheme->getBinningMethod() )
-   {
-   case BINNING_CARTESIAN_PRODUCT:
-      generateCartesianProductPairs( m_couplingScheme );
-      break;
-   case BINNING_GRID:
-      // The spatial grid is templated on the dimension
+   case BINNING_BVH:
+      // The BVH is templated on the dimension and execution space
       switch( dim )
       {
       case 2:
-         m_gridSearch2D->findInterfacePairs();
+         switch(cs->getExecutionMode())
+         {
+            case(ExecutionMode::Sequential):
+               m_search = new BvhSearch<2, axom::SEQ_EXEC>(m_coupling_scheme);
+               break;
+            #ifdef TRIBOL_USE_OPENMP
+            case(ExecutionMode::OpenMP):  // This causes compiler to hang
+               //SLIC_ERROR("Unsupported execution mode: " << parameters.exec_mode );
+               m_search = new BvhSearch<2, axom::OMP_EXEC>(m_coupling_scheme);
+               break;
+            #endif
+            #ifdef TRIBOL_USE_CUDA
```

Comment:
Do we need the cuda and hip ifdefs to be inside a TRIBOL_USE_RAJA? Maybe that is more of a consistency thing with other parts of the code?